src/ApiCore/OperationResponse.php

```diff
@@ -133,40 +164,68 @@ public function getName()
      * Poll the server in a loop until the operation is complete.
      *
      * Return true if the operation completed, otherwise return false. If the
-     * $options['maxPollingDuration'] setting is not set (or set <= 0.0) then
+     * $options['totalPollTimeoutMillis'] setting is not set (or set <= 0.0) then
      * pollUntilComplete will continue polling until the operation completes,
      * and therefore will always return true.
      *
      * @param array $options {
      *                       Options for configuring the polling behaviour.
      *
-     *     @type float $pollingIntervalSeconds The polling interval to use, in seconds.
-     *                                          Default: 1.0
-     *     @type float $maxPollingDurationSeconds The maximum amount of time to continue polling.
-     *                                            Default: 0.0 (no maximum)
+     *     @type float $initialPollDelayMillis The initial polling interval to use, in seconds.
+     *     @type float $pollDelayMultiplier    Multiplier applied to the polling interval on each retry.
+     *     @type float $maxPollDelayMillis     The maximum polling interval to use, in seconds.
+     *     @type float $totalPollTimeoutMillis The maximum amount of time to continue polling.
      * }
      * @throws ApiException If an API call fails.
+     * @throws ValidationException
      * @return bool Indicates if the operation completed.
      */
     public function pollUntilComplete($options = [])
     {
-        $defaultPollSettings = [
-            'pollingIntervalSeconds' => $this::DEFAULT_POLLING_INTERVAL,
-            'maxPollingDurationSeconds' => 0.0,
-        ];
-        $pollSettings = array_merge($defaultPollSettings, $options);
+        $pollSettings = array_merge($this->defaultPollSettings, $options);
+
+        $currentPollDelayMillis = $pollSettings['initialPollDelayMillis'];
```

Comment:
These settings seem similar to those used by the retry settings. Do you think it may make sense to have an implementation that accepts these settings and performs the repeated call logic on our behalf? It could be something we use to eventually replace the classes like [`Google\Cloud\Core\ExponentialBackoff`](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/Core/src/ExponentialBackoff.php) and could use in our system tests amongst other places.

Reply:
Yes, they are similar. I had a look at trying to combine them, but it is harder than it first seems, because the retrying middleware class needs to retry over promises, so its logic is quite different.

I have moved the polling logic into a trait - let me know if you think this is an improvement or not. I think that this could be used in place of the exponential backoff class - or else, it could supplement it and replace the polling/sleep logic.